src/Language/PureScript/Environment.hs

```diff
@@ -129,10 +130,24 @@ makeTypeClassData
   -> [FunctionalDependency]
   -> Bool
   -> TypeClassData
-makeTypeClassData args m s deps = TypeClassData args m s deps determinedArgs coveringSets
+makeTypeClassData args m s deps = TypeClassData args m' s deps determinedArgs coveringSets
   where
     ( determinedArgs, coveringSets ) = computeCoveringSets (length args) deps
 
+    coveringSets' = S.toList coveringSets
+
+    m' = map (\(a, b) -> (a, b, addVtaInfo b)) m
+    
+    addVtaInfo :: SourceType -> Maybe (S.Set (NEL.NonEmpty Int))
+    addVtaInfo memberTy = do
+      let mentionedArgIndexes = S.fromList (mapMaybe argToIndex $ freeTypeVariables memberTy)
+      let leftovers = map (`S.difference` mentionedArgIndexes) coveringSets'
+      let vtaRequiredArgs = S.fromList $ mapMaybe (NEL.nonEmpty . S.toList) leftovers
+      if S.null vtaRequiredArgs then Nothing else Just vtaRequiredArgs
```

Comment:
Oh dear, this doesn't look right.

Let me walk through this:
* `mentionedArgIndexes` is a `Set Int` of the (indices of the) arguments that participate in this member's type
* `leftovers` is a `[Set Int]` made from taking each covering set and subtracting `mentionedArgIndexes`, so there is one set in `leftovers` for every covering set
* `vtaRequiredArgs` is a `Set (NonEmpty Int)` comprising only the non-empty sets in `leftovers`—but if there are *any* empty sets in `leftovers`, then there is no need for VTA info, because there is a way to disambiguate the type arguments based on `mentionedArgIndexes` via the covering set corresponding to the empty `leftovers` element
* the result is `Nothing` only if `vtaRequiredArgs` is empty—which only happens if *all* of the sets in `leftovers` are empty

I think what you want instead is

```suggestion
      S.fromList <$> traverse (NEL.nonEmpty . S.toList) leftovers
```

And ideally one of the test cases would cover this by having:
* multiple distinct covering sets
* a member that mentions all of the arguments in one, but not all, of the covering sets
* expected error without the VTA hint

Reply:
Oo... good catch!

Reply:
All of the current tests still pass when this line is reverted. Could you add a case like the following:

```purs
class TestClass a b | a -> b, b -> a where
  testMethod :: Maybe a -> Int

test :: Int
test = testMethod Nothing
```

The `NoInstanceFound` should have an ‘unknown type variables’ hint but not a VTA hint.